Make regenerating story trees run as a job with a job status reporter API

nick-funk opened the pull request:
## What does this PR do?

Moves the regenerate story trees functionality into a job processor queue.

Adds an api router at `<CORAL_URL>/api/jobStatus` that can be called with the `name`: `"regenerateStoryTrees` and return the progress of the story tree generation.

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds optional `disableCommenting` and `disableCommentingMessage` fields to the `RegenerateStoryTreesInput` type.

## Does this PR introduce any new environment variables or feature flags? 

No new environment variables.

## If any indexes were added, were they added to `INDEXES.md`?

No new indexes.

## How do I test this PR?

- Start up Coral
- Create some stories or delete the `tree` field off of some stories
- Call the regenerate story trees graphql query to `<CORAL_URL>/api/graphql` with:
     ```
     mutation {
       regenerateStoryTrees(input: {
         clientMutationId: "1",
         disableCommenting: true,
         disableCommentingMessage: "doing story tree regeneration, closing comments while I think really hard"
       }) {
         accepted
         jobID
       }
     }
     ```
- Remember the `jobID` this returns, you'll need it in a moment
- Wait a minute or so, the job takes a bit to get going and start posting progress since it does a count to check how many stories it's going to update so it can report accurate progress
- Call `<CORAL_URL>/api/jobStatus` authenticated with an admin user token with the payload until you see that the percentage is `100`
    - NOTE: this is rate limited, so you can only call it twice a minute
    ```
    {
       "name": "regenerateStoryTrees"
       "jobID": "<YOUR_JOB_ID>"
    }
    ```
- Check your stories, see that they have correct `tree` fields generated for all stories under the tenant.
 
## How do we deploy this PR?

Requires the same deployment process at the 7.1.0 story tree migration since this is a fix to remedy the story tree generation issues in 7.1.0.

nick-funk committed:
Use job processing to regenerate story trees

nick-funk committed:
Create job status api endpoint

Currently only returns results for the `regenerateStoryTrees` job
if the appropriate name is supplied.

nick-funk committed:
Hook up jobStatus api router with necessary middleware

nick-funk committed:
Start regenerateStoryTrees job task so that it will run

Also set its timeout to 8 hours for total runtime, we don't
want regenerateStoryTrees to run for longer than one evening.
If clients need more time to regenerate story trees, they
likely should look into archiving some of their older stories.

nick-funk committed:
Make closing commenting an optional param for story tree regeneration

Automates the closing of the stories during regen if wanted

nick-funk committed:
Disable Z_KEY/COMMENT_SEEN if available when disabling comments during story tree gen

marcushaddon reviewed (approved):
These changes look good to me, also tested locally and works as advertised!

nick-funk committed:
Fix batch bucket leak in story tree regeneration

Count was not being set to 0 during batch generation causing
the system to write 1 story at a time after the first batch
generation.

nick-funk committed:
Account for null stories in regenerate story trees bulk op

nick-funk committed:
warn on errors but continue processing

outputs the story ID's that weren't processed so we can debug
them if necessary.

nick-funk committed:
Handle super deep comment streams when generating trees

nick-funk committed:
Use redis to store/retrieve job status progress

nick-funk committed:
Fix off by one error in generate story flatten function

nick-funk committed:
Process 15 story trees per regeneration batch

nick-funk committed:
Properly handle super deep replies edge case

Just stop the recursion, grab all the offspring of the root (ancestor)
comment and return it, no need to go down and do further recursion since
all the comments are in sorted order.

nick-funk committed:
Make comment insertion handle super deep replies properly

nick-funk committed:
When encountering an archived story, generate a tree for it

If the tree already exists, we carry on and just use it for finding
the next unseen comment

nick-funk committed:
Show started and ended time in jobStatus handler result
